### ISSUE TITLE
Gramine installation through apt should check for existing Gramine installations built from source

### ISSUE THREAD

**Issue by marcelamelara** [gramineproject/gramine#901]
## Description of the problem

There is currently no straightforward way to cleanly uninstall old versions of Gramine that were installed from source without the original Gramine source tree. In other words, ` sudo ninja -C build uninstall` does not work in cases where the Gramine source tree is removed from the machine, and a new ninja build is configured at a later point with a newer version of the Gramine source tree. 

Manually removing Gramine binaries and lib files from `/usr/local` directories may lead to an incomplete/unclean uninstall.

## Steps to reproduce
```
cd gramine
meson setup build/ --buildtype=debug -Ddirect=enabled -Dsgx=enabled -Dsgx_driver=upstream
ninja -C build/
sudo ninja -C build/ install

// at a later time
rm -rf gramine
git clone https://github.com/gramineproject/gramine.git
cd gramine
meson setup build/ --buildtype=debug -Ddirect=enabled -Dsgx=enabled -Dsgx_driver=upstream
ninja -C build/
sudo ninja -C build uninstall 
```

## Expected results
In cases where Gramine is installed through apt, the Gramine installation config should check for existing/older versions of Gramine in /usr/local and (at the very least) warn the user. 

## Actual results
Accidental side-by-side/conflicting installations of Gramine in `/usr` and `/usr/local` if Gramine is ever installed via apt without properly uninstalling Gramine built from source first.

## Gramine commit hash

Tested with Gramine source tree at f7995b6
Gramine version installed via apt: 1.2

**Comment by boryspoplawski** [comment]
1. I don't think what you are asking for is even possible. You can install Gramine at virtually any place in file system (any prefix).
2. If you have a custom build I recommend having a custom prefix. Things like `/opt/gramine/` as prefix is probably what most people would use.

tl;dr I believe everything works as it should.

**Comment by dimakuv** [comment]
If I understand correctly, `apt install` installs Gramine (on Debian/Ubuntu) at `/usr/`. Whereas `sudo ninja -C build/ install` installs Gramine at `/usr/local/`.

I guess this is where the problem of incompatible versions stems from. If this is indeed so, I agree that this is a nightmare to debug/realize.

We can't be the only project who has such problems... What do other projects do in this situation? @woju

**Comment by boryspoplawski** [comment]
`sudo ninja install` installs gramine at `--prefix=XYZ` defined at `meson setup`. `XYZ` can be anything you want, `/usr/local/` is just it's default value, but can be changed.

**Comment by dimakuv** [comment]
@boryspoplawski If I understand correctly, `meson setup` defaults to the prefix of `/usr/local/`. Which sounds fair, and I don't think we should change it.

At the same time, Debian/Ubuntu packages install under `/usr/`. You can see it in Woju's PR: https://github.com/gramineproject/gramine/pull/512/files (find the line `MESONARGS := --prefix=/usr`). This `/usr/` prefix is I guess something that Debian/Ubuntu defaults to and mandates from software.

So there are two ways to correctly install Gramine (default `meson setup` and then `ninja install` vs `apt install`), and the two ways install Gramine into two different directories (`/usr/local/` vs `/usr/`).

My question is generic -- how do normal users deal with such situations? It feels absolutely normal to try both ways (like, you first install Gramine via `apt install`, then try the latest version via `ninja install`). And in this case, if you don't know about `/usr/` vs `/usr/local/` default installation policies, then you are stuck with some garbage state and need to ask us or google extensively.

Is this just one of those things in the Linux world where developers/packagers say "Well, that's how it works, get used to random rules like this"?

**Comment by boryspoplawski** [comment]
As I said, when you build Gramine manually you are supposed to tweak `--prefix` to your needs.

> It feels absolutely normal to try both ways (like, you first install Gramine via apt install, then try the latest version via ninja install)

It does not to me. If you want two versions of the same software installed, you should keep them separated.
Also I believe you shouldn't install stuff manually into `/usr/` (or whatever your distro uses) - this is the place for stuff coming from your system directly (here apt), which correctly tracks all dependencies and conflicts.

Also I believe doing `sudo ninja uninstall` **from a different** `meson setup`-ed directory is not supposed ever to work and is asking for trouble.

Note than I'm in no means a sysadmin or anything related, these are my private opinions backed on, well, really nothing.

**Comment by marcelamelara** [comment]
Thank you both @dimakuv and @boryspoplawski for your input.

I agree, for folks like me doing research with Gramine, switching between `ninja install` and `apt install` versions is not atypical. But I understand if this isn't the _most_ typical use case for Gramine, too.

That said, I understand that my problem was ultimately the fact that _I_ removed the original `meson setup`-ed build directory before doing a `ninja uninstall`. I only attempted to do the `ninja uninstall` in a different source tree _because_ I did not want to simply manually search and `rm` the Gramine files in `/usr/local`, because that is also asking for trouble. And, for a security-critical application like Gramine, that's not what the solution should be in any case. 

More generally, my concerns come from the SGX attestation validation standpoint. Side-by-side installations of Gramine can lead to validation errors due to confusion about the libpal.so used to generate and validate the MRENCLAVE. But for folks who _want_ side-by-side installations of Gramine-SGX, it would be helpful to at least point out the potential for these types of SGX/security-related issues in documentation, even if it's not possible to check for existing installations during a `ninja install` or an `apt install`.

**Comment by dimakuv** [comment]
I agree with @marcelamelara. Gramine should do something about this. Maybe Gramine could at least probe `which gramine-direct` (or some Gramine tool that always exists, no matter what backend you install) during `meson setup` and give a huge warning `Gramine already installed under ..., are you sure you want to continue with installation? N/y`.

**Comment by woju** [comment]
> We can't be the only project who has such problems... What do other projects do in this situation?

Nothing at all. Building and installing from source is error-prone esp. wrt uninstalling and user is expected to know what s/he is doing. The exact scenario described in the original issue is impossible to correct for (you can't reliably uninstall things if you don't have original builddir, the new builddir might be configured differently and has different set of installed files). Even if you have original builddir, `uninstall` might be broken and wouldn't undo some changes. Some of those are not obvious if should be done at all. (Do you want to retain app-generated configs/dbs or not? `apt-get` has two different subcommands for those two cases). For example, our `ninja uninstall` is broken, because it won't undo symlinks (grep mesonfiles for `meson.add_install_script`). We don't care though, because it's a general and well-known problem that you can't uninstall things compiled from source. If you want reliable uninstalling, you should install/uninstall using distro's packaging tools only (apt+deb or dnf+rpm). All other solutions are in principle unreliable.

> I did not want to simply manually search and rm the Gramine files in /usr/local, because that is also asking for trouble

The expectation is that you should know what you're doing if you install from source. It installs things in `prefix` as you've configured it. The default `/usr/local` is a sane thing to do to `just work` if you didn't think about it (it does, because `/usr/local/bin` is in `$PATH` on most distros), and also because `/usr/local` is generally safe to mess with, because there are no system-level things there, they're all in `/usr`. You can get non-working gramine, but you won't brick your system. Sorry to be blunt, but this all is basic Linux admin skills, and we're generally not in business of educating people in basics.

> for a security-critical application like Gramine, that's not what the solution should be in any case.

Umm, what security? If you want production-level security, you need to install binary packages, because they're signed directly. Our sources are currently unsigned, so unless you trust GitHub and WebPKI (an unlikely model given the interest in SGX), installing from sources is insecure.

**Comment by dimakuv** [comment]
Thanks @woju. I'm surprised by the current state of affairs (`it's a general and well-known problem that you can't uninstall things compiled from source`). Well, if there is no common solution to this, and if other projects do `nothing at all`, I guess we can close this issue as wontfix.

> Sorry to be blunt, but this all is basic Linux admin skills, and we're generally not in business of educating people in basics.

I'd argue that UNIX created this mess, and learning 100500 less-than-obvious rules to acquire basic Linux admin skills is just terrible.

**Comment by woju** [comment]
> Well, if there is no common solution to this

The common solution really is to install from packages. It just works, incl. uninstall.

> I'd argue that UNIX created this mess, and learning 100500 less-than-obvious rules to acquire basic Linux admin skills is just terrible.

Sadly, the problem of uninstallation is not limited to *nix (exhibit A: `uninstall.exe`, exhibit B: `regedit32.exe`). *nix is unique only because here installation from source is actually a thing, and in traditionally handcrafted makefiles the developers are responsible for sane `make uninstall`. But developers rarely uninstall their own software, so make target gets probably the least love of all of them.

**Comment by marcelamelara** [comment]
> Sorry to be blunt, but this all is basic Linux admin skills, and we're generally not in business of educating people in basics.

I totally disagree that this is simply _a basic Linux admin issue_. Sure, I already readily admitted my basic mistake, AND this does not mean that you can't find ways to do better for other folks. Because side-by-side installations do have **security** implications, the absolute least you can do is point this out in documentation, because **that** is non-obvious to newbie SGX users. If you want Gramine to be adopted by more folks, especially in the context of SGX, you need to do better.

> if there is no common solution to this, and if other projects do nothing at all, I guess we can close this issue as wontfix.

Understand there is little motivation to fix this, thank you @dimakuv for enabling the discussion.
